azure/services/scalesets/scalesets.go

```diff
@@ -276,6 +278,18 @@ func (s *Service) patchVMSSIfNeeded(ctx context.Context, infraVMSS *azure.VMSS)
 		return nil, errors.Wrap(err, "failed to calculate maxSurge")
 	}
 
+	replicaseManagedByAutoscaler := s.Scope.HasReplicasExternallyManaged(ctx)
+
+	hasCustomDataChanges, err := s.Scope.HasBootstrapDataChanges(ctx)
+	if err != nil {
+		log.V(4).Error(err, "unable to calculate custom data hash, ignoring")
+	}
+	if hasCustomDataChanges {
+		log.V(4).Info("custom data changed")
+	} else {
+		log.V(4).Info("custom data unchanged")
+	}
+
 	hasModelChanges := hasModelModifyingDifferences(infraVMSS, vmss)
```

Comment:
Have you checked if we get customData back from the existing VMSS? if so we might not even need to store the custom data hash since we would be able to just compare the custom data we have with the existing VMSS custom data

Reply:
We do return the customData from the fetchedVMSS state under the getVirtualMachineScaleSet() function, however the SDKtoVMSS converters remove the OSProfile data in which its stored. If this is the preferred method I can update the converters and types to facilitate this comparison rather than storing a hash.

Reply:
yes I think that would be a cleaner implementation and it can rely on the actual state of the world instead needing to store the hash in the API. We can modify the `SDKtoVMSS` converters helper function easily to do what we need since it's part of the CAPZ code.

Reply:
Ok I think I've gone through and solidified the logic to use the SDKtoVMSS data, just waiting for the tests to green light and I'll mark this as resolved.